dataclassy/dataclass.py

```diff
@@ -51,16 +51,19 @@ def __new__(mcs, name, bases, dict_, **kwargs):
         # determine whether a post-initialisation method was defined by the user
         init = dict_.get('__init__')
         post_init = (type(init) is Function and not hasattr(init, '__dataclass__')
-                     or type(dict_.get('__post_init__')) is Function)
+                     or any(type(b.get('__post_init__')) is Function
+                            for b in dataclass_bases + [dict_]))
 
         # fill out this class' dict and store defaults, annotations and decorator options for future subclasses
+        all_defaults.pop('__classcell__', None) #do not mess with dict_['__classcell__'] in case it exists in one of dataclass_bases but not in dict_
```

Comment:
I'm worried that this is a CPython [implementation detail](https://docs.python.org/3/reference/datamodel.html#creating-the-class-object) and is not portable to PyPy.

Reply:
I guess it's OK even if it doesn't exist on PyPy but why is it needed?

Reply:
To be honest I don't know exactly how `__classcell__` works, but if you pass the wrong `__classcell__` to `type.__init__` it messes with the interpretation of `super()` in subclasses.

Reply:
I think a better approach might be to only put variables in `__defaults__` if they're also in `__annotations__`.  That's what I did in my other PR, but it can also be done without the MRO change.  That way `__classcell__` is fixed automatically.

Reply:
Yes, this sounds like a good idea. I'll merge the changes we already have before changing that since it might also affect logic around slots.